### ISSUE TITLE
Does gocql implement async writes? (Cont)

### ISSUE THREAD

**Issue by rodgarrison** [gocql/gocql#1309]
This is a followup to https://github.com/gocql/gocql/issues/781

If we have 10,000,000 records i.e. an enormous number for insertion then since `gocql` doesn't support asynch calls in the API, is the recommendation then to set an upper limit N where N is the number of goroutines and channels and proceed like this?

1. make N channels
2. make N goroutines which each routine listens to 1 of the N channels
3. in the main thread, iterate through the 10,000,000 records passing each record into one of the N channels say in a round robin fashion probably blocking when the channels are full

And have the goroutines insert the record and pass back error/success. We have no concerns about having processes with all the threads GoLang already comes with plus the N additional threads and channels? We don't care about the GC and related overhead? This is considered production worthy?

If the application says has 10 record types it needs to insert, each with a similarly large number of records, and channels operate only on a single type, this solution becomes a bit more complicated. Each routine would need more complexity to deal with the various types (preferred I guess) or the initial N increases to 10*N if we keep code paths separated. That's potentially a lot of additional memory, overhead.

**Comment by alourie** [comment]
@rodgarrison what is the source of the records? I'm not sure if the solution to writing this amount of data lies within the driver here. Not saying it shouldn't be able to do it....just wondering if it's the best approach.

**Comment by rodgarrison** [comment]
The records are being to delivered to the DB writer via Kafka. Now in fairness to `gosql` is not like the 10,000,000 enormous records for insertion are just sitting their arriving out of thin air. So the truer description would be the DB write rate has to equal or exceed the input rate. Nonetheless the application would still require the N channels, and N go routines to handle the throughput. (This application is application will indeed expect work loads of 1000s-to-low-10000s per second). So the overhead of those channels and goroutines seems to be an unavoidable overhead.

**Comment by alourie** [comment]
@rodgarrison well, in this case, you're basically writing a Kafka consumer that needs to ingest that data. I don't see any way to avoid the overhead except for writing a client with tolerable utilisation levels and then run such clients on many machines to support the required accumulated consumption rate. It's quite a common scenario for reading from Kafka topics anyway, and I assume more reliable as well, as running many instances of the topic consumer will be a bit more fault tolerant.

In short, I don't see a way to avoid the overhead and personally would look into running this via many small instances running vs one big application.

**Comment by martin-sucha** [comment]
Gocql can execute multiple queries concurrently, but you need a goroutine for each execution where you call gocql to execute the query. There is no need to have N channels, one channel with N goroutines executing the queries is enough (where N is number of concurrently running queries).

Closing this issue as everything is working as intended, providing synchronous APIs is the recommended way in Go. If somebody thinks we need an asynchronous API because of performance reasons, please provide some performance benchmark/data that will demonstrate the need.
